lib/templateFactory.js

```diff
@@ -80,32 +74,67 @@ class TemplateFactory extends BaseFactory {
      */
-    getTemplate(config) {
-        // get all templates with the same name
-        return super.list({ params: { name: config.name } })
+    getTemplate(fullTemplateName) {
+        const [, templateName, versionOrTag] = TEMPLATE_NAME_REGEX.exec(fullTemplateName);
+        const isExactVersion = EXACT_VERSION_REGEX.exec(versionOrTag);
+        const isVersion = VERSION_REGEX.exec(versionOrTag);
+
+        if (isExactVersion) {
+            // Get a template using the exact template version
+            return super.get({
+                name: templateName,
+                version: versionOrTag
+            });
+        }
+
+        // If tag is passed in, get the version from the tag
+        if (versionOrTag && !isVersion) {
+            // Lazy load factory dependency to prevent circular dependency issues
+            // eslint-disable-next-line global-require
+            const TemplateTagFactory = require('./templateTagFactory');
+            const templateTagFactory = TemplateTagFactory.getInstance();
+
+            // Get a template tag
+            return templateTagFactory.get({
+                name: templateName,
+                tag: versionOrTag
+            })
+            .then((templateTag) => {
+                // Return null if no template tag exists
+                if (!templateTag) {
+                    return null;
+                }
+
+                // Get a template using the exact template version
+                return super.get({
+                    name: templateName,
+                    version: templateTag.version
+                });
+            });
+        }
+
+        // Get all templates with the same name
+        return super.list({ params: { name: templateName } })
             .then((templates) => {
-                // get templates that have version prefix as config.version
+                // If no version provided, return the most recently published template
+                if (!versionOrTag) {
+                    return templates[0];
+                }
+
+                // Get templates that have versions beginning with the version given
                 const filtered = templates.filter(template =>
-                    template.version.startsWith(config.version));
+                    template.version.startsWith(versionOrTag));
 
                 // Sort templates by descending order
                 filtered.sort((a, b) => compareVersions(b.version, a.version));
 
-                // If no config.label, return latest one
-                if (filtered.length > 0 && !config.label) {
-                    return filtered[0];
-                }
-
-                // Loop through filtered to find the first match for config.label
-                return filtered.find(template => template.labels.includes(config.label));
+                // Return first filtered template or null if none
+                return filtered[0] ? filtered[0] : null;
```

Comment:
I think `filtered[0] || null` is a bit cleaner. Other than that everything looks good!